contracts/lotto/src/contract.rs

```diff
@@ -0,0 +1,1154 @@
+    }
+    // Increment total deposit
+    let balance: Coin = info
+        .funds
+        .iter()
+        .filter(|coin| coin.denom == ticket_price.denom)
+        .last()
+        .unwrap()
+        .clone();
+
+    lotto.balance += balance.amount;
+    // Add participant address
+    lotto.participants.push(info.sender.clone());
+
+    // Save the state & updated config escrow balance
+    LOTTOS.save(deps.storage, lotto_id, &lotto)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "participate")
+        .add_attribute("sender", info.sender.as_ref())
+        .add_attribute("new_balance", lotto.balance.to_string()))
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    callback: NoisCallback,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // callback should only be allowed to be called by the proxy contract
+    // otherwise anyone can cut the randomness workflow and cheat the randomness by sending the randomness directly to this contract
+    ensure_eq!(
+        info.sender,
+        config.nois_proxy,
+        ContractError::UnauthorizedReceive
+    );
+    let randomness: [u8; 32] = callback
+        .randomness
+        .to_array()
+        .map_err(|_| ContractError::InvalidRandomness)?;
+
+    // extract lotto nonce
+    let job_id = callback.job_id;
+    let lotto_nonce: u64 = job_id
+        .strip_prefix("lotto-")
+        .expect("Strange, how is the job-id not prefixed with lotto-")
+        .parse()
+        .unwrap(); //Needs to check that the received nonce is a number
+
+    // Make sure the lotto nonce is valid
+    let lotto = LOTTOS.load(deps.storage, lotto_nonce)?;
+    assert!(lotto.winners.is_none(), "Strange, there's already winners");
+    let participants = lotto.participants;
+
+    let winners = nois::pick(
+        randomness,
+        lotto.number_of_winners as usize,
+        participants.clone(),
+    );
+
+    if winners.is_empty() {
+        return Err(ContractError::NoDepositors {});
+    }
+
+    let amount_creator = get_percentage(lotto.balance, config.creator_commission_percent);
+    let amount_protocol = get_percentage(lotto.balance, config.protocol_commission_percent);
+    let winnable_amount = lotto.balance - amount_creator - amount_protocol;
+
+    let mut recipients_list_amounts: Vec<(Addr, Uint128)> = vec![];
+    let mut recipients_list_amounts_total: Uint128 = Uint128::new(0);
+    for recipient in lotto.recipients_list.clone() {
+        let recipient_amount = get_percentage(winnable_amount, recipient.1);
+        recipients_list_amounts.push((recipient.0, recipient_amount));
+        recipients_list_amounts_total += recipient_amount;
+    }
+
+    let prize_amount = winnable_amount - recipients_list_amounts_total;
+    let amount_winner = prize_amount.multiply_ratio(
+        Uint128::new(1),
+        Uint128::new(lotto.number_of_winners as u128),
+    );
```

Comment:
Often this amount will be lower than the `prize_amount` due to rounding. E.g. prize_amount = 10 and `lotto.number_of_winners = 3`, then this is 10/3 = 3 and you distribute 3*3 = 9.

I suggest pulling out this logic and unit testing it.

Reply:
hmm true, in practical situations the Uint is probably gonna be in the order of millions so the rounding would have much less impact. But in principle this needs a better handling